### PR TITLE
Allow Ethernet Connections (arduino)

### PR DESCRIPTION
**Pull request by chrisflesher**
## Pull Request Details
This is the arduino side needed to support Ethernet and WiFi101 connections.

## Breaking Changes
There are no breaking changes, this PR just adds an example .ino files to add Ethernet and WiFi support. I've only tested the wifi101 demo, I think the ethernet example is so similar to the WiFi it *should* work but haven't tested yet.

## Other Relevant Information
Added matching code to [`jfjlaros/arduino-simple-rpc` PR #4](https://github.com/jfjlaros/arduino-simple-rpc/pull/4). The basic strategy is to use the pyserial `serial_for_url` function call instead of `Serial` constructor. This allows you to set the device name to be something like `socket://192.168.1.151:10000` instead of `/dev/ttyACM0`. I was able to get WiFi working using the command `simple_rpc list -d socket://192.168.1.151:10000`

### CODE REVIEW

File: examples/ethernet/ethernet.ino [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  connectEthernet();
+  printEthernetStatus(Serial);
+  server.begin();
+}
+
+void loop(void) {
+  EthernetClient client = server.available();
+  if (client) {
+    while (client.connected()) {
+      if (client.available()) {
```
Comment (jfjlaros) [inline examples/ethernet/ethernet.ino:40]: I wonder if this line is needed, in `interface.tcc` on line 113, we seem to do the same thing.

File: examples/ethernet/ethernet.ino [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  if (Ethernet.linkStatus() == LinkOFF) {
+    print.println("Ethernet cable is not connected.");
+  }
+}
+
+void setup(void) {
+  Serial.begin(9600);
+  Ethernet.begin(mac, ip);
+  connectEthernet();
+  printEthernetStatus(Serial);
```
Comment (jfjlaros) [inline examples/ethernet/ethernet.ino:32]: Using the serial device might break RPC over serial.

File: examples/ethernet/ethernet.ino [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+byte ping(byte data) {
+  return data;
+}
+
+void connectEthernet() {
+  if (Ethernet.hardwareStatus() == EthernetNoHardware) {
+    while (true) {
+      delay(1);  // do nothing, no point running without Ethernet hardware
+    }
```
Comment (jfjlaros) [inline examples/ethernet/ethernet.ino:16]: What about this one:

    while (Ethernet.hardwareStatus() == EthernetNoHardware) {
      delay(1);
    }

**Review by jfjlaros — COMMENTED** [review]
I added a few suggestions.

**Review by chrisflesher — COMMENTED** [review]
@jfjlaros thanks for reviewing, this is ready for another pass.

**Comment by chrisflesher** [comment]
There still seems to be some issue with this approach I haven't fully debugged. When I call `simple_rpc call ping 1 -d socket://192.168.1.50:10000` it returns a zero. I found that the python side is writing a `1` (on line 166 of `simple_rpc.py`) and that the arduino ping function receives a `0`...

**Comment by jfjlaros** [comment]
Maybe it is a timing issue? Do you know whether the `Stream` read and write calls are blocking or non-blocking?

**Comment by chrisflesher** [comment]
I think they are usually non-blocking. For my work we often use rs485
connections and have to use a flush and microsecond delay to get
transmission to work sometimes.

**Comment by jfjlaros** [comment]
I only had blocking I/O in mind when writing this library, using non-blocking I/O will probably break functionality at higher speeds.

**Comment by chrisflesher** [comment]
Hmm... I would have thought `Serial` would be non-blocking as well...

https://www.arduino.cc/reference/en/language/functions/communication/serial/flush/

**Comment by chrisflesher** [comment]
I tried adding a `io.flush()` to `rpcPrint` and `rpcWrite`, it didn't help, still debugging...

**Comment by chrisflesher** [comment]
Here is a log of the bytes I am getting with the function `simple_rpc call ping 1 -d socket://192.168.1.50:10000`:
```
rpcRead: FF 
rpcPrint: 73 69 6D 70 6C 65 52 50 43 
rpcPrint: 3C 
rpcPrint: 49 
rpcPrint: 42 
rpcPrint: 3A 
rpcPrint: 20 
rpcPrint: 42 
rpcPrint: 69 
rpcPrint: 3A 
rpcPrint: 20 
rpcPrint: 69 
rpcPrint: 3A 
rpcPrint: 20 
rpcPrint: 42 
rpcPrint: 4C 
rpcPrint: 3A 
rpcPrint: 28 
rpcPrint: 63 
rpcPrint: 28 
rpcPrint: 69 
rpcPrint: 63 
rpcPrint: 29 
rpcPrint: 29 
rpcPrint: 3A 
rpcPrint: 20 
rpcPrint: 28 
rpcPrint: 63 
rpcPrint: 28 
rpcPrint: 69 
rpcPrint: 63 
rpcPrint: 29 
rpcPrint: 29 
rpcPrint: 5B 
rpcPrint: 66 
rpcPrint: 5D 
rpcPrint: 3A 
rpcPrint: 20 
rpcPrint: 5B 
rpcPrint: 69 
rpcPrint: 5D 
rpcPrint: 5B 
rpcPrint: 66 
rpcPrint: 5D 
rpcPrint: 3A 
rpcPrint: 20 
rpcPrint: 5B 
rpcPrint: 69 
rpcPrint: 5D 
rpcRead: 00 
rpcRead: 00 
rpcWrite: 00 
rpcRead: 01 
rpcRead: 00 20001101 00 20001188 
rpcWrite: 01 01 00 06 
```

It seems like the problem is the second to last `rpcWrite 00` immediately before it gets the data from `rpcRead 01`?

**Comment by jfjlaros** [comment]
Which sketch are you using?

**Comment by chrisflesher** [comment]
Am using the wifi101 sketch with some extra Serial.print and io.flush
statements in the .tcc files not shown in this pr for debugging.

**Comment by jfjlaros** [comment]
This is the output I get when doing more or less the same thing using the serial interface:

    >>> from serial import Serial
    >>> c = Serial('/dev/ttyACM0')
    >>> c.write(bytes([0, 1]))
    2
    >>> c.read()
    b'\x01'
    >>> c.read()

The last instruction blocks because there is no more data.

Could you share the output of a similar exercise for the ethernet interface?

**Comment by chrisflesher** [comment]
Getting the same thing (both manually and run as a script):
```
>>> from serial import serial_for_url
>>> c = serial_for_url('socket://192.168.1.50:10000')
>>> c.write(bytes([0, 1]))
2
>>> print(c.read())
b'\x01'
>>> print(c.read())
```

**Comment by chrisflesher** [comment]
However the same example split into two individual writes seems to cause problems:
```
>>> from serial import serial_for_url
>>> c = serial_for_url('socket://192.168.1.50:10000')
>>> c.write(bytes([0]))
1
>>> c.write(bytes([1]))
1
>>> print(c.read())
b'\x01'
>>> print(c.read())
b'\x02'
```

**Comment by jfjlaros** [comment]
What happens if you start reading immediately? E.g.,

    >>> from serial import serial_for_url
    >>> c = serial_for_url('socket://192.168.1.50:10000')
    >>> print(c.read())

If this and successive reads returns anything, my guess would be that `read()` is nonblocking.

**Comment by jfjlaros** [comment]
As a test, could you add the following line to the top of `StreamIO::read` in `plugins/stream/io.cpp` and repeat your test with two individual writes?

    while (!available());

**Comment by chrisflesher** [comment]
With this example
```
>>> from serial import serial_for_url
>>> c = serial_for_url('socket://192.168.1.50:10000')
>>> print(c.read())
```
It waits indefinitely.

I'm using the `WiFiClient` which is a `Stream` object but doesn't use `plugins/stream/io.cpp` so couldn't test that.

I was able to get it working by changing the `rpcRead` function to be blocking:

```
/*! \ingroup read
 * Read a value from an Input / output object.
 *
 * \param io Input / output object.
 * \param data Data.
 */
template <class I, class T>
void rpcRead(I& io, T* data) {
  while (io.available() < sizeof(T)) {
    delay(1);
  }
  io.read((uint8_t*)data, sizeof(T));
}
```

I'm not sure if this is a great idea long term or not since it introduces the chance for the microcontroller to get locked up when waiting for data to arrive.

**Comment by chrisflesher** [comment]
The WiFi connection seems to be working now.

**Comment by jfjlaros** [comment]
Good to hear that you found a workaround. I agree that this might not be a long term solution, so could you also try this fix in `plugins/stream/io.cpp`?

    void StreamIO::begin(Stream& ss) {
      _ss = &ss;
      _ss->setTimeout(1000);
    }

If I understand it correctly, this should make the read blocking but it will give up after 1 second.

**Comment by chrisflesher** [comment]
Ok, here is a new suggestion. I found the Arduino `Stream` class already has a default timeout of 1 second and we can use the `readBytes` function to implement the timeout. This was working before for `Serial` because it was already using `readBytes` under the hood (in `plugins/StreamIO/io.cpp`).

My suggestion is to rename all `io.read` calls to `io.readBytes`, this allows you to use `Stream` objects directly with the `interface` function call instead of going through the plugins wrapper and take advantage of the timeout. I think this simplifies the code a bit but maybe there is something I'm not envisioning with the plugins, if you want to keep the plugin style of doing things we could wrap the `WiFiClient` in the `StreamIO` plugin.

**Comment by jfjlaros** [comment]
Yes, that sounds like a good solution.

So now we can simply pass either `Serial` or `WiFiClient` to the interface function, making the whole plugin structure redundant?

**Comment by chrisflesher** [comment]
Yes, you should be able to just supply `Serial` to the interface function call now.

**Comment by jfjlaros** [comment]
Great, that would save quite some code and template magic.

I guess that if we do need a wrapper for some other protocol, we can make it a subclass of `Stream` as well.

**Comment by chrisflesher** [comment]
Well I tried running the `Serial` demo and it didn't work, will try to look at it some more...

**Comment by chrisflesher** [comment]
Ok, this is working now passing `Serial` directly to `interface`.

**Comment by jfjlaros** [comment]
Thank you, I confirmed this works for me as well. Nice work.

I will update the test fixture before I will do the merge.

**Pull request merged**